src/agent.py

```diff
@@ -0,0 +1,100 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Jenkins agent relation observer."""
+import logging
+import typing
+
+from ops.charm import CharmBase, RelationJoinedEvent
+from ops.framework import Object
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus
+
+import jenkins
+from state import State
+
+logger = logging.getLogger(__name__)
+
+
+class AgentRelationData(typing.TypedDict):
+    """Relation data required for adding the Jenkins agent.
+
+    Attrs:
+        url: The Jenkins server url.
+        secret: The secret for agent node.
+    """
+
+    url: str
+    secret: str
+
+
+class Observer(Object):
+    """The Jenkins agent relation observer."""
+
+    def __init__(self, charm: CharmBase, state: State):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+            state: The charm state.
+        """
+        super().__init__(charm, "agent-observer")
+        self.charm = charm
+        self.state = state
+
+        charm.framework.observe(charm.on["agent"].relation_joined, self._on_agent_relation_joined)
+
+    @property
+    def _jenkins_container(self) -> Container:
+        """The Jenkins workload container."""
+        return self.charm.unit.get_container(self.state.jenkins_service_name)
+
+    def _on_agent_relation_joined(self, event: RelationJoinedEvent) -> None:
+        """Handle agent relation joined event.
+
+        Args:
+            event: The event fired from an agent joining the relationship.
+        """
+        if not (binding := self.model.get_binding("juju-info")):
```

Comment:
Under what conditions would the binding not be available?

Reply:
This was to remove the None type, updated to assertion, thank you!